src/tuner/algo/allreduce/ring.h

```diff
@@ -0,0 +1,311 @@
+ */
+static inline unsigned long long get_ring_chunk_steps(int proto)
+{
+    if      (NCCL_PROTO_LL     == proto) { return NCCL_OFI_TUNER_RING_LL_CHUNK_STEPS; }
+    else if (NCCL_PROTO_LL128  == proto) { return NCCL_OFI_TUNER_RING_LL128_CHUNK_STEPS; }
+    else if (NCCL_PROTO_SIMPLE == proto) { return NCCL_OFI_TUNER_RING_SIMPLE_CHUNK_STEPS; }
+    assert(false);
+    __builtin_unreachable();
+}
+
+/**
+ * @brief Get the step size for a given protocol.
+ *
+ * @param proto The protocol (NCCL_PROTO_SIMPLE, NCCL_PROTO_LL, or NCCL_PROTO_LL128).
+ * @return The step size for the given protocol.
+ */
+static inline unsigned long long get_ring_stepsize(int proto)
+{
+    return get_buffsize(proto) / NCCL_OFI_TUNER_NCCL_STEPS;
+}
+
+/**
+ * @brief Get the chunk size for a given protocol.
+ *
+ * @param proto The protocol (NCCL_PROTO_SIMPLE, NCCL_PROTO_LL, or NCCL_PROTO_LL128).
+ * @return The step size for the given protocol.
+ */
+static inline unsigned long long get_ring_chunksize(int proto)
+{
+    return (get_ring_stepsize(proto) * get_ring_chunk_steps(proto) / get_overhead_denominator(proto)) * get_overhead_numerator(proto);
+}
+
+/**
+ * @brief Get the chunk window in bytes.
+ *
+ * @param proto The protocol (NCCL_PROTO_SIMPLE, NCCL_PROTO_LL, or NCCL_PROTO_LL128).
+ * @return The number of chunks for the given protocol.
+ */
+static inline unsigned long long get_ring_chunk_window_in_bytes(int proto)
+{
+	return get_ring_chunk_steps(proto)*get_buffsize(proto)*get_overhead_numerator(proto)/get_overhead_denominator(proto)/NCCL_OFI_TUNER_NCCL_STEPS;
+}
+
+/**
+ * @brief Get the size of the ring "window" for a given protocol, number of channels, and number of
+ *        ranks.
+ *
+ * @param num_channels The number of channels.
+ * @param num_ranks The number of ranks.
+ * @param proto The protocol (NCCL_PROTO_SIMPLE, NCCL_PROTO_LL, or NCCL_PROTO_LL128).
+ * @return The size of the ring window.
+ */
+static inline double get_ring_size_window(unsigned long long num_channels, unsigned long long num_ranks, int proto)
+{
+    return num_channels * num_ranks * get_ring_chunk_window_in_bytes(proto);
+}
+
+/**
+ * @brief Get the number of "windows" required for a given data size, protocol, number of channels,
+ *        and number of ranks. This is a term we made up for our tuner, NCCL may use a different
+ *        name to describe this.
+ *
+ * @param nbytes The data size in bytes.
+ * @param num_channels The number of channels.
+ * @param num_ranks The number of ranks.
+ * @param proto The protocol (NCCL_PROTO_SIMPLE, NCCL_PROTO_LL, or NCCL_PROTO_LL128).
+ * @return The number of windows required.
+ */
+static inline int get_ring_num_windows(unsigned long long nbytes, unsigned long long num_channels, unsigned long long num_ranks, int proto)
+{
+    const double size_window = 1.0*get_ring_size_window(num_channels, num_ranks, proto);
+    const double iters = ceil(nbytes / size_window);
+    int int_iters = iters;
+    return int_iters + 1;
```

Comment:
why +1?
For example, when nbytes = 1, this function should return 1. right?

Reply:
Sure, please see: 189b013 `fixup! tuner: better ring rank/msize binning` and review the changes in the decisions.